src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp

```diff
@@ -0,0 +1,214 @@
+    assert(Universe::is_module_initialized(), "sanity");
+  }
+
+  const char* category_name = AOTClassLinker::class_category_name(class_category);
+  switch (class_category) {
+  case AOTLinkedClassCategory::BOOT1:
+    load_classes_impl(class_category, table->boot(), category_name, loader, CHECK);
+    break;
+
+  case AOTLinkedClassCategory::BOOT2:
+    load_classes_impl(class_category, table->boot2(), category_name, loader, CHECK);
+    break;
+
+  case AOTLinkedClassCategory::PLATFORM:
+    {
+      initiate_loading(THREAD, category_name, loader, table->boot());
+      initiate_loading(THREAD, category_name, loader, table->boot2());
+      load_classes_impl(class_category, table->platform(), category_name, loader, CHECK);
+    }
+    break;
+  case AOTLinkedClassCategory::APP:
+    {
+      initiate_loading(THREAD, category_name, loader, table->boot());
+      initiate_loading(THREAD, category_name, loader, table->boot2());
+      initiate_loading(THREAD, category_name, loader, table->platform());
+      load_classes_impl(class_category, table->app(), category_name, loader, CHECK);
+    }
+    break;
+  case AOTLinkedClassCategory::UNREGISTERED:
+    ShouldNotReachHere(); // Currently aot-linked classes are not supported for this category.
+    break;
+  }
+}
+
+void AOTLinkedClassBulkLoader::load_classes_impl(AOTLinkedClassCategory class_category, Array<InstanceKlass*>* classes,
+                                                 const char* category_name, Handle loader, TRAPS) {
+  if (classes == nullptr) {
+    return;
+  }
+
+  ClassLoaderData* loader_data = ClassLoaderData::class_loader_data(loader());
+
+  for (int i = 0; i < classes->length(); i++) {
+    InstanceKlass* ik = classes->at(i);
+    if (log_is_enabled(Info, cds, aot, load)) {
+      ResourceMark rm(THREAD);
+      log_info(cds, aot, load)("%-5s %s%s%s", category_name, ik->external_name(),
+                               ik->is_loaded() ? " (already loaded)" : "",
+                               ik->is_hidden() ? " (hidden)" : "");
+    }
+
+    if (!ik->is_loaded()) {
+      if (ik->is_hidden()) {
+        // TODO: AOTClassLinking is not implemented for hidden class until JDK-8293336
+        ShouldNotReachHere();
+      } else {
+        InstanceKlass* actual;
+        if (loader_data == ClassLoaderData::the_null_class_loader_data()) {
+          actual = SystemDictionary::load_instance_class(ik->name(), loader, CHECK);
+        } else {
+          actual = SystemDictionaryShared::find_or_load_shared_class(ik->name(), loader, CHECK);
+        }
+
+        if (actual != ik) {
+          ResourceMark rm(THREAD);
+          log_error(cds)("Unable to resolve %s class from CDS archive: %s", category_name, ik->external_name());
+          log_error(cds)("Expected: " INTPTR_FORMAT ", actual: " INTPTR_FORMAT, p2i(ik), p2i(actual));
+          log_error(cds)("JVMTI class retransformation is not supported when archive was generated with -XX:+AOTClassLinking.");
```

Comment:
Nit: use a `logStream` instead of the three separate calls.

Reply:
Why?

Reply:
@stefank pointed out to me that wouldn't actually be atomic unless you used `NonInterleavingLogStream`.